File: backend/analyser.py
"""
analyser.py — Hybrid SBERT + Keyword Topic Matcher (multi-label)
=================================================================
Maps each question to one or more syllabus topics using:
  final_score = 0.7 × sbert_similarity + 0.3 × keyword_overlap

Returns multi-label assignments with confidence scores, so one question
can belong to multiple topics (e.g., "CFG and cyclomatic complexity").
"""

import re
import os
from collections import defaultdict
from typing import Optional

import numpy as np

# Detect environment and decide SBERT usage before importing heavy deps.
_USE_SBERT_ENV = os.getenv("USE_SBERT", "").strip().lower()
_RUNNING_ON_RENDER = bool(os.getenv("RENDER")) or bool(os.getenv("RENDER_EXTERNAL_URL"))

# Default to disabling SBERT on Render free tier to avoid OOM, unless explicitly enabled.
if _USE_SBERT_ENV:
    USE_SBERT = _USE_SBERT_ENV in {"1", "true", "yes", "y", "on"}
else:
    USE_SBERT = False if _RUNNING_ON_RENDER else True

# Optional SBERT import (can be heavy due to torch)
if USE_SBERT:
    try:
        from sentence_transformers import SentenceTransformer
        _SBERT_MODEL = None  # lazy-loaded
        HAS_SBERT = True
    except ImportError:
        HAS_SBERT = False
        _SBERT_MODEL = None
else:
    HAS_SBERT = False
    _SBERT_MODEL = None

from sklearn.metrics.pairwise import cosine_similarity


# ═══════════════════════════════════════════════════════════════════════════
# Model loading (singleton, lazy)
# ═══════════════════════════════════════════════════════════════════════════

def _get_model():
    global _SBERT_MODEL
    if _SBERT_MODEL is None and HAS_SBERT:
        print("[analyser] Loading SBERT model (first run only)...")
        _SBERT_MODEL = SentenceTransformer("all-MiniLM-L6-v2")
        print("[analyser] Model loaded.")
    return _SBERT_MODEL


# ═══════════════════════════════════════════════════════════════════════════
# Keyword overlap scorer (TF-IDF style bag-of-words)
# ═══════════════════════════════════════════════════════════════════════════

_STOP_WORDS = {
    "the", "a", "an", "and", "or", "of", "in", "to", "for", "with",
    "is", "are", "was", "be", "by", "on", "at", "from", "that", "this",
    "what", "how", "why", "which", "any", "all", "its", "their", "it",
    "explain", "describe", "define", "discuss", "compare", "state",
    "elaborate", "illustrate", "identify", "give", "list", "show",
    "find", "compute", "calculate", "write", "draw", "outline",
}


def _tokenize(text: str) -> set[str]:
    """Lowercase tokenize, drop stop words and short tokens."""
    tokens = re.findall(r'[a-z0-9]+', text.lower())
    return {t for t in tokens if t not in _STOP_WORDS and len(t) > 2}


def keyword_overlap_score(question: str, topic: str) -> float:
    """
    Jaccard-style overlap between question tokens and topic tokens.
    Returns [0, 1].
    """
    q_tokens = _tokenize(question)
    t_tokens = _tokenize(topic)
    if not q_tokens or not t_tokens:
        return 0.0
    intersection = q_tokens & t_tokens
    # Weighted: intersect / topic_tokens (recall-biased, so topic match matters more)
    return len(intersection) / len(t_tokens)


# ═══════════════════════════════════════════════════════════════════════════
# SBERT similarity scorer
# ═══════════════════════════════════════════════════════════════════════════

def sbert_similarity_batch(
    questions: list[str],
    topics: list[str],
    model,
) -> np.ndarray:
    """
    Compute cosine similarity matrix: shape (n_questions, n_topics).
    """
    q_embs = model.encode(questions, show_progress_bar=False, normalize_embeddings=True)
    t_embs = model.encode(topics,    show_progress_bar=False, normalize_embeddings=True)
    return cosine_similarity(q_embs, t_embs)  # (n_q, n_t)


# ═══════════════════════════════════════════════════════════════════════════
# Hybrid scorer
# ═══════════════════════════════════════════════════════════════════════════

SBERT_WEIGHT   = 0.7
KEYWORD_WEIGHT = 0.3
TOP_K_TOPICS   = 3    # max topics per question
MIN_CONFIDENCE = 0.25  # minimum score to assign a topic
\n

def score_questions_against_syllabus(
    questions: list[dict],   # list of {"text": ..., "q_num": ..., "module": ...}
    syllabus_flat: list[dict],  # [{"topic": ..., "module": ..., "module_title": ...}]
    use_sbert: bool = True,
) -> list[dict]:
    """
    For each question, assign up to TOP_K_TOPICS with confidence scores.

    Returns list of enriched question dicts:
    {
      ...original question fields...,
      "topics": [
        {"name": "Huffman Coding", "module": 2, "module_title": "...", "score": 0.87},
        {"name": "Prefix Codes",   "module": 2, "module_title": "...", "score": 0.61},
      ],
      "primary_module": 2,   # module of highest-scoring topic
    }
    """
    enriched = []
    
    # We load SBERT if requested
    model = _get_model() if (use_sbert and HAS_SBERT and USE_SBERT) else None

    for qi, q in enumerate(questions):
        # 1. Hard-map the module using parser
        target_module = q.get("module", 0)

        # 2. Extract ONLY topics for that target module
        # If parsing failed and module is 0, we'll fall back to searching all modules.
        # But if module is known (1-5), STRICTLY limit the search space to that module.
        if target_module > 0:
            local_topics = [t for t in syllabus_flat if t["module"] == target_module]
        else:
            local_topics = syllabus_flat
            
        if not local_topics:
            enriched.append({**q, "topics": [], "primary_module": target_module, "is_uncertain": True})
            continue

        local_texts = [t["topic"] for t in local_topics]

        # 3. Compute NLP similarity strictly against the local topics
        if model:
            sbert_scores = sbert_similarity_batch([q["text"]], local_texts, model)[0]
        else:
            sbert_scores = None

        topic_scores = []
        for ti, topic_info in enumerate(local_topics):
            kw_score = keyword_overlap_score(q["text"], topic_info["topic"])
            sb_score = float(sbert_scores[ti]) if sbert_scores is not None else kw_score
            
            hybrid = SBERT_WEIGHT * sb_score + KEYWORD_WEIGHT * kw_score
            topic_scores.append((hybrid, topic_info))

        # Sort descending, take top-K
        topic_scores.sort(key=lambda x: x[0], reverse=True)
        assigned = []
        
        # Check highest score for uncertainty flag
        max_score = topic_scores[0][0] if topic_scores else 0.0
        is_uncertain = max_score < 0.4
        
        for score, topic_info in topic_scores[:TOP_K_TOPICS]:
            if score >= MIN_CONFIDENCE:
                assigned.append({
                    "name":         topic_info["topic"],
                    "module":       topic_info["module"],
                    "module_title": topic_info["module_title"],
                    "score":        round(score, 4),
                })

        primary_module = assigned[0]["module"] if assigned else target_module

        enriched.append({
            **q,
            "topics":         assigned,
            "primary_module": primary_module,
            "is_uncertain":   is_uncertain
        })

    return enriched


# ═══════════════════════════════════════════════════════════════════════════
# Frequency aggregation
# ═══════════════════════════════════════════════════════════════════════════

def compute_topic_frequency(
    all_enriched_questions: list[dict],
    syllabus_flat: list[dict] = None
) -> dict:
    """
    Aggregate topic frequency across all questions from all papers.
    Each topic gets:
      - count: total weighted occurrences (sum of scores)
      - raw_count: number of questions mentioning this topic
      - module: which module this topic belongs to
      - questions: list of questions referencing this topic
    """
    freq: dict[str, dict] = defaultdict(lambda: {
        "count": 0.0,
        "raw_count": 0,
        "module": 0,
        "module_title": "",
        "questions": [],
    })

    if syllabus_flat:
        for st in syllabus_flat:
            key = st["topic"]
            freq[key]["module"] = st["module"]
            freq[key]["module_title"] = st.get("module_title", "")
            freq[key]["count"] = 0.0
            freq[key]["raw_count"] = 0
            freq[key]["questions"] = []

    for q in all_enriched_questions:
        for t in q.get("topics", []):
            key = t["name"]
            freq[key]["count"]        += t["score"]
            freq[key]["raw_count"]    += 1
            freq[key]["module"]        = t["module"]
            freq[key]["module_title"]  = t["module_title"]
            freq[key]["questions"].append({
                "text":   q["text"],
                "q_num":  q.get("q_num"),
                "source": q.get("source", ""),
                "score":  t["score"],
            })

    return dict(freq)


def get_ranked_topics(freq: dict, module: Optional[int] = None) -> list[dict]:
    """
    Return topics sorted by raw_count descending (most repeated first).
    Optionally filter by module.
    """
    items = [
        {"topic": k, **v}
        for k, v in freq.items()
        if (module is None or v["module"] == module)
    ]
    items.sort(key=lambda x: (x["raw_count"], x["count"]), reverse=True)
    return items

File: frontend/src/lib/analyser.js
/**
 * analyser.js — JS port of backend/analyser.py
 * ===============================================
 * Replaces SBERT with Jaccard keyword overlap + synonym expansion.
 *
 * Scoring:
 *   score = keywordOverlapScore(question, topic) ∈ [0, 1]
 *
 * Topic assignment per question:
 *   - Strict module isolation (only match topics from the question's own module)
 *   - Keep top-3 matches where score ≥ MIN_CONFIDENCE (0.40)
 *   - Label each match: Strong (≥0.70), Medium (0.40–0.69)
 */

// ─────────────────────────────────────────────────────────────────────────────
// Stop words
// ─────────────────────────────────────────────────────────────────────────────

const STOP_WORDS = new Set([
  'the','a','an','and','or','of','in','to','for','with','is','are','was','be',
  'by','on','at','from','that','this','what','how','why','which','any','all',
  'its','their','it','explain','describe','define','discuss','compare','state',
  'elaborate','illustrate','identify','give','list','show','find','compute',
  'calculate','write','draw','outline','using','use','brief','short','note',
  'difference','between','example','examples','write','discuss',
]);


// ─────────────────────────────────────────────────────────────────────────────
// Synonym dictionary  (canonical term → array of alternate forms)
// ─────────────────────────────────────────────────────────────────────────────

const SYNONYMS = {
  'cyclomatic complexity':    ['mccabe complexity', 'cyclomatic number', 'mccabe metric'],
  'distributed system':       ['distributed computing', 'distributed environment', 'distributed processing'],
  'huffman coding':           ['huffman tree', 'huffman encoding', 'huffman algorithm'],
  'finite automata':          ['finite state machine', 'fsa', 'dfa', 'nfa', 'deterministic automata'],
  'context free grammar':     ['cfg', 'pushdown automata', 'pda', 'context-free language'],
  'time complexity':          ['big o', 'asymptotic analysis', 'asymptotic complexity', 'order of complexity'],
  'space complexity':         ['memory complexity', 'auxiliary space'],
  'dynamic programming':      ['dp', 'memoization', 'tabulation'],
  'minimum spanning tree':    ['mst', 'prims algorithm', 'kruskals algorithm', 'spanning tree'],
  'shortest path':            ['dijkstra', 'bellman ford', 'single source shortest path'],
  'sorting':                  ['merge sort', 'quick sort', 'heap sort', 'bubble sort', 'insertion sort'],
  'binary search tree':       ['bst', 'binary tree', 'search tree'],
  'regular expression':       ['regex', 'regular language', 'regular grammar'],
  'turing machine':           ['tm', 'universal turing machine'],
  'mutual exclusion':         ['mutex', 'critical section', 'semaphore'],
  'deadlock':                 ['deadlock detection', 'deadlock prevention', 'deadlock avoidance'],
  'virtual memory':           ['paging', 'page fault', 'demand paging', 'page replacement'],
  'process scheduling':       ['cpu scheduling', 'round robin', 'sjf', 'fcfs', 'priority scheduling'],
  'cache memory':             ['cache', 'cache hit', 'cache miss', 'locality of reference'],
  'instruction pipeline':     ['pipelining', 'pipeline stages', 'data hazard', 'control hazard'],
};

// Build reverse lookup: alias → canonical
const _ALIAS_MAP = new Map();
for (const [canonical, aliases] of Object.entries(SYNONYMS)) {
  for (const alias of aliases) {
    _ALIAS_MAP.set(alias, canonical);
  }
}


// ─────────────────────────────────────────────────────────────────────────────
// Tokeniser
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Lowercase, remove stop words, keep tokens ≥ 3 chars.
 * Also expands synonyms: if a token matches an alias, the canonical form is added.
 * @param {string} text
 * @returns {Set<string>}
 */
export function tokenize(text) {
  const raw    = text.toLowerCase().match(/[a-z0-9]+/g) || [];
  const tokens = new Set();

  for (const t of raw) {
    if (t.length < 3 || STOP_WORDS.has(t)) continue;
    tokens.add(t);
    // Expand aliases — single-word check
    if (_ALIAS_MAP.has(t)) tokens.add(_ALIAS_MAP.get(t).split(' ')[0]);
  }

  // Multi-word alias check (bigrams / trigrams)
  const words = text.toLowerCase().replace(/[^a-z0-9 ]/g, ' ').split(/\s+/);
  for (let n = 2; n <= 3; n++) {
    for (let i = 0; i <= words.length - n; i++) {
      const phrase = words.slice(i, i + n).join(' ').trim();
      if (_ALIAS_MAP.has(phrase)) {
        // Add each word of the canonical term
        for (const w of _ALIAS_MAP.get(phrase).split(' ')) {
          if (w.length >= 3) tokens.add(w);
        }
      }
    }
  }

  return tokens;
}


// ─────────────────────────────────────────────────────────────────────────────
// Jaccard recall-biased overlap score
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Score ∈ [0, 1].  Recall-biased: |intersection| / |topic_tokens|
 * so a short, specific topic phrase is easier to match than a broad one.
 * @param {string} questionText
 * @param {string} topicStr
 * @returns {number}
 */
export function keywordOverlapScore(questionText, topicStr) {
  const qTokens = tokenize(questionText);
  const tTokens = tokenize(topicStr);
  if (qTokens.size === 0 || tTokens.size === 0) return 0;

  let intersect = 0;
  for (const t of tTokens) {
    if (qTokens.has(t)) intersect++;
  }
  return intersect / tTokens.size;
}


// ─────────────────────────────────────────────────────────────────────────────
// Confidence labelling
// ─────────────────────────────────────────────────────────────────────────────

const MIN_CONFIDENCE = 0.40;
const TOP_K          = 3;

function confidenceLabel(score) {
  if (score >= 0.70) return 'Strong';
  if (score >= 0.40) return 'Medium';
  return 'Weak';
}


// ─────────────────────────────────────────────────────────────────────────────
// Core scorer
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Enrich each question with matched topics from its own module.
 *
 * @param {Array<{text, q_num, module, source}>} questions   Flat list of all questions
 * @param {Array<{topic, module, module_title}>}  syllabusFlat
 * @returns {Array}  questions with added { topics[], primary_module, is_uncertain }
 */
export function scoreQuestionsAgainstSyllabus(questions, syllabusFlat) {
  return questions.map(q => {
    const targetModule = q.module || 0;

    // Strict module isolation
    const localTopics = targetModule > 0
      ? syllabusFlat.filter(t => t.module === targetModule)
      : syllabusFlat;

    if (!localTopics.length) {
      return { ...q, topics: [], primary_module: targetModule, is_uncertain: true };
    }

    // Score each topic
    const scored = localTopics.map(topicInfo => ({
      score:     keywordOverlapScore(q.text, topicInfo.topic),
      topicInfo,
    }));

    // Sort descending, take top-K above threshold
    scored.sort((a, b) => b.score - a.score);
    const topScore = scored[0]?.score ?? 0;

    const assigned = scored
      .slice(0, TOP_K)
      .filter(({ score }) => score >= MIN_CONFIDENCE)
      .map(({ score, topicInfo }) => ({
        name:         topicInfo.topic,
        module:       topicInfo.module,
        module_title: topicInfo.module_title,
        score:        Math.round(score * 10000) / 10000,
        label:        confidenceLabel(score),
      }));

    return {
      ...q,
      topics:         assigned,
      primary_module: assigned[0]?.module ?? targetModule,
      is_uncertain:   topScore < 0.40,
    };
  });
}


// ─────────────────────────────────────────────────────────────────────────────
// Frequency aggregation
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Count how often each topic appears across all questions.
 *
 * @param {Array}  allEnrichedQuestions
 * @param {Array}  syllabusFlat  (to pre-seed all topics with zero counts)
 * @returns {Object}  { topicName: { count, raw_count, module, module_title, questions[] } }
 */
export function computeTopicFrequency(allEnrichedQuestions, syllabusFlat = []) {
  const freq = {};

  // Pre-seed every topic so it appears in charts even with 0 occurrences
  for (const st of syllabusFlat) {
    if (!freq[st.topic]) {
      freq[st.topic] = {
        count:        0,
        raw_count:    0,
        module:       st.module,
        module_title: st.module_title,
        questions:    [],
      };
    }
  }

  for (const q of allEnrichedQuestions) {
    for (const t of (q.topics || [])) {
      if (!freq[t.name]) {
        freq[t.name] = { count: 0, raw_count: 0, module: t.module, module_title: t.module_title, questions: [] };
      }
      freq[t.name].count    += t.score;
      freq[t.name].raw_count += 1;
      freq[t.name].module        = t.module;
      freq[t.name].module_title  = t.module_title;
      freq[t.name].questions.push({
        text:   q.text,
        q_num:  q.q_num,
        source: q.source || '',
        score:  t.score,
      });
    }
  }

  return freq;
}

/**
 * Return topics sorted by raw_count descending (optionally filtered by module).
 * @param {Object}   freq
 * @param {number|null} module
 * @returns {Array}
 */
export function getRankedTopics(freq, module = null) {
  const items = Object.entries(freq)
    .filter(([, v]) => module === null || v.module === module)
    .map(([topic, v]) => ({ topic, ...v }));

  items.sort((a, b) => b.raw_count - a.raw_count || b.count - a.count);
  return items;
}
